postgresql/src/storages/postgres/tests/result_set_pgtest.cpp

```diff
@@ -117,4 +117,28 @@ UTEST_P(PostgreConnection, ResultTraverseBackward) {
   EXPECT_EQ(5, num);
 }
 
+UTEST_P(PostgreConnection, ResultAsOptionalSingleRow) {
+  CheckConnection(GetConn());
+
+  pg::ResultSet res{nullptr};
+  UEXPECT_NO_THROW(res = GetConn()->Execute("select * from "
+                                            "(values (4, 3)) as data"));
+  ASSERT_EQ(1, res.Size());
+
+  UEXPECT_NO_THROW(res.AsOptionalSingleRow<int>());
+
+}
+
+UTEST_P(PostgreConnection, ResultAsOptionalSingleRowThrows) {
+  CheckConnection(GetConn());
+
+  pg::ResultSet res{nullptr};
+  UEXPECT_NO_THROW(res = GetConn()->Execute("select * from "
+                                            "(values (4, 3), (2, 1)) as data"));
+  ASSERT_EQ(2, res.Size());
+
+  UEXPECT_THROW(res.AsOptionalSingleRow<int>(), pg::NonSingleRowResultSet);
+}
+
+
```

Comment:
You are missing the case when AsOptionalSingleRow actually returns null.
To simulate a query that does not result in any rows I would use something like `select * from generate_series(0, -1)`